src/server/qgsfcgiserverresponse.cpp

```diff
@@ -98,33 +115,59 @@ void QgsSocketMonitoringThread::run( )
   }
 
 #if defined(Q_OS_UNIX) && !defined(Q_OS_ANDROID)
+#ifdef QGISDEBUG
+  const pid_t threadId = gettid();
+#endif
+
+  mIsResponseFinished.store( false );
   char c;
-  while ( !*mIsResponseFinished )
+  while ( !mIsResponseFinished.load() )
   {
     const ssize_t x = recv( mIpcFd, &c, 1, MSG_PEEK | MSG_DONTWAIT ); // see https://stackoverflow.com/a/12402596
-    if ( x < 0 )
+    if ( x != 0 )
     {
       // Ie. we are still connected but we have an 'error' as there is nothing to read
-      QgsDebugMsgLevel( QStringLiteral( "FCGIServer: remote socket still connected. errno: %1" ).arg( errno ), 5 );
+      QgsDebugMsgLevel( QStringLiteral( "FCGIServer %1: remote socket still connected. errno: %2, x: %3" )
+                        .arg( threadId ).arg( errno ).arg( x ), 5 );
     }
     else
     {
-      // socket closed, nothing can be read
-      QgsDebugMsgLevel( QStringLiteral( "FCGIServer: remote socket has been closed! errno: %1" ).arg( errno ), 2 );
-      mFeedback->cancel();
-      break;
+      // double check...
+      ssize_t x2 = 0;
+      for ( int i = 0; !mIsResponseFinished.load() && x2 == 0 && i < 10; i++ )
+      {
+        x2 = recv( mIpcFd, &c, 1, MSG_PEEK | MSG_DONTWAIT ); // see https://stackoverflow.com/a/12402596
+        std::this_thread::sleep_for( 50ms );
```

Comment:
Why waiting ? Is it require for synchronization?

Reply:
I will remove this extra check!